services/com.mde-network.ep.toolfunctions.xtext/src/main/java/com/mdenetnetwork/ep/toolfunctions/xtext/XtextTool.java

```diff
@@ -0,0 +1,227 @@
+import org.apache.hc.client5.http.HttpHostConnectException;
+import org.apache.hc.client5.http.entity.mime.MultipartEntityBuilder;
+import org.apache.hc.core5.http.ContentType;
+import org.apache.hc.core5.http.HttpEntity;
+import org.apache.hc.client5.http.entity.mime.HttpMultipartMode;
+import org.apache.hc.client5.http.fluent.Request;
+
+public class XtextTool  { 
+
+	static final String PROJECT_PATH = "./xtext-project/" ;
+	static final String PROJECT_FILENAME = "xtext-project.zip" ;
+	static final String PROJECT_FILE_PATH = "./" + PROJECT_FILENAME ;
+	static final String ES_JSON_RESPONSE_FIELD = "editorUrl";
+	static final String ES_JSON_RESPONSE_FIELD_STATUS = "editorStatusUrl";
+	static final int FILE_BUFFER_SIZE = 1024;
+	
+	public XtextTool() {
+		
+		// Register resource factories
+		Resource.Factory.Registry.INSTANCE.getExtensionToFactoryMap().put("ecore", new EcoreResourceFactoryImpl());
+	}
+
+
+	public void run(String languageName, String baseName, String extension, String grammar, List<ProjectFile> projectFiles, OutputStream outputStream, JsonObject response) throws Exception {
+
+
+		
+		// Create Xtext project with web editor enabled
+		File projectFolder = new File(PROJECT_PATH);
+		
+		String[] args = new String[] {
+				"-targetDir=" + projectFolder.getAbsolutePath(),
+				"-languageName="+languageName,
+				"-baseName="+baseName,
+				"-extension="+extension,
+				"-enabledProjects=UIProject,GenericIDE,RuntimeTest,UITest,Web",
+				"-buildSystem=MAVEN",
+				"-sourceLayout=PLAIN",
+				"-languageServer=FATJAR",
+				"-xtextVersion=2.30.0",
+				"-projectLayout=FLAT",
+				"-javaVersion=JAVA11"
+			};
+
+		
+		CliProjectsCreatorMain.main(args);
+		
+		
+		
+		
+		// Add activity files
+		if (grammar != null) {
+			final String grammarSrcPath = languageName.replace('.', '/');
+			final String xtextGrammarPath= PROJECT_PATH + baseName + "/src/" + grammarSrcPath  + ".xtext";
+			recreateGrammarFile(xtextGrammarPath, grammar);
+		}
+		
+		//TODO add activity files
+		
+		
+		// Compress files for transfer to build server
+		
+		ZipOutputStream projectFileStream = new ZipOutputStream( new FileOutputStream(PROJECT_FILE_PATH) );
+		
+		compressDirectory(projectFolder, "", projectFileStream);
+		 
+		projectFileStream.flush();
+		projectFileStream.close();
+		 
+		
+		// Send to the build server
+		
+		final String EDITOR_SERVER_URL= "http://127.0.0.1:10001/xtext/upload";  // TODO set URL via environment variable 
+		 
+		 try {	 
+			 postProjectToUrl(new File(PROJECT_FILE_PATH), EDITOR_SERVER_URL, response);
+			 
+		 } catch (HttpHostConnectException e)  {
+		 
+			 System.err.println("Failed to connect to editor server: " + EDITOR_SERVER_URL);
+			 
+		}finally {
+			 // Cleanup created files
+			 deleteDir(projectFolder);
+			 new File(PROJECT_FILE_PATH).delete();
+		 }
+
+	}
+
+	
+	public void convertXToY(String X,  OutputStream outputStream, JsonObject response) throws Exception {
```

Comment:
What's the purpose of this function? It doesn't seem to be doing anything at the moment. Can it be removed?

Reply:
Agreed I'll remove it.